File: services/narrative-director/src/utils/converters.js
import { EmotionPrediction, PlayerBehaviorData } from '../proto/emotion_pb.js'

export function toEmotionPredictionObject({ dominantEmotion, confidence, timestampMs }) {
  const p = new EmotionPrediction()
  if (dominantEmotion) p.setDominantEmotion(dominantEmotion)
  if (confidence != null) p.setConfidence(confidence)
  if (timestampMs != null) p.setTimestampMs(timestampMs)
  return p.toObject()
}

export function toPlayerBehaviorProto(payload) {
  const d = new PlayerBehaviorData()
  if (payload.sessionId) d.setSessionId(payload.sessionId)
  if (payload.timestampMs != null) d.setTimestampMs(payload.timestampMs)
  if (payload.decisionLatencyMs != null) d.setDecisionLatencyMs(payload.decisionLatencyMs)
  if (payload.actionFrequency != null) d.setActionFrequency(payload.actionFrequency)
  if (Array.isArray(payload.recentChoices)) d.setRecentChoicesList(payload.recentChoices)
  return d
}

File: services/emotion-inference/src/converters.py
from proto import emotion_pb2

def to_proto_player_behavior(data):
    ts = int((data.get("timestamp_ms") or 0))
    dl = float((data.get("decision_latency_ms") or 0.0))
    af = int((data.get("action_frequency") or 0))
    rc = [int(x) for x in (data.get("recent_choices") or [])]
    ctx = {str(k): float(v) for k, v in (data.get("context") or {}).items()}
    return emotion_pb2.PlayerBehaviorData(
        session_id=str(data.get("session_id")),
        timestamp_ms=ts,
        decision_latency_ms=dl,
        action_frequency=af,
        recent_choices=rc,
        context_data=ctx,
    )

def proto_to_response(pred: emotion_pb2.EmotionPrediction):
    return {
        "dominant_emotion": pred.dominant_emotion,
        "confidence": pred.confidence,
        "all_scores": dict(pred.all_scores),
        "timestamp_ms": pred.timestamp_ms,
    }
